src/main/sensors/rpm_filter.c

```diff
@@ -0,0 +1,202 @@
+    config->gyro_rpm_notch_q = 50;
+
+    config->dterm_rpm_notch_harmonics = 1;
+    config->dterm_rpm_notch_min = 100;
+    config->dterm_rpm_notch_q = 50;
+}
+
+static void rpmNotchFilterInit(rpmNotchFilter_t* filter, int harmonics, int minHz, int q, float looptime) 
+{
+    filter->harmonics = harmonics;
+    filter->minHz = minHz;
+    filter->q = q / 10.0f;
+
+    for (int axis = 0; axis < XYZ_AXIS_COUNT; axis++) {
+        for (int motor = 0; motor < getMotorCount(); motor++) {
+            for (int i = 0; i < harmonics; i++) {
+                biquadFilterInit(
+                    &filter->notch[axis][motor][i], minHz * i, looptime, filter->q, FILTER_NOTCH);
+            }
+        }
+    }
+}
+
+void rpmFilterInit(const rpmFilterConfig_t *config)
+{
+    pidLooptime = gyro.targetLooptime * pidConfig()->pid_process_denom;
+    rpmNotchFilterInit(&gyroFilter, config->gyro_rpm_notch_harmonics,
+                       config->gyro_rpm_notch_min, config->gyro_rpm_notch_q, gyro.targetLooptime);  
+    rpmNotchFilterInit(&dtermFilter, config->dterm_rpm_notch_harmonics,
+                       config->dterm_rpm_notch_min, config->dterm_rpm_notch_q, pidLooptime);
+
+    for (int i = 0; i < getMotorCount(); i++) {
+        pt1FilterInit(&rpmFilters[i], pt1FilterGain(RPM_MOTOR_FILTER_CUTOFF, gyro.targetLooptime));
+    }
+    
+    const float secondsPerMinute = 60.0f;
+    const float erpmPerLsb = 100.0f;
+    erpmToHz = erpmPerLsb / secondsPerMinute  / (motorConfig()->motorPoleCount / 2.0f);
+
+    const float minUpdateT = 0.001f;
+    const float loopIterationsPerUpdate = minUpdateT / (pidLooptime * 1e-6f);
+    numberFilters = getMotorCount() * (gyroFilter.harmonics + dtermFilter.harmonics);
+    const float filtersPerLoopIteration = numberFilters / loopIterationsPerUpdate;
+    filterUpdatesPerIteration = rintf(filtersPerLoopIteration + 0.49f);
+}
+
+static float applyFilter(rpmNotchFilter_t* filter, int axis, float value)
+{
+    for (int motor = 0; motor < getMotorCount(); motor++) {
+        for (int i = 0; i < filter->harmonics; i++) {
+            value = biquadFilterApplyDF1(&filter->notch[axis][motor][i], value);
+        }
+    }
+    return value;
+}
+
+float rpmFilterGyro(int axis, float value)
+{
+    if (gyroFilter.harmonics == 0) {
+        return value;
+    } else {
+        return applyFilter(&gyroFilter, axis, value);
+    }
+}
+
+float rpmFilterDterm(int axis, float value)
+{
+    if (dtermFilter.harmonics == 0) {
+        return value;
+    } else {
+        return applyFilter(&dtermFilter, axis, value);
+    }
+}
+static float motorFrequency[MAX_SUPPORTED_MOTORS];
+
+void rpmFilterUpdate()
+{
+    if (gyroFilter.harmonics == 0 && dtermFilter.harmonics == 0) {
+        return;
+    }
+    
+    for (int motor = 0; motor < getMotorCount(); motor++) {
+        filteredMotorErpm[motor] = pt1FilterApply(&rpmFilters[motor], getDshotTelemetry(motor));
+        DEBUG_SET(DEBUG_RPM_FILTER, motor, motorFrequency[motor]);
```

Comment:
The filters are not updated with old data. Whenever rpmFilterUpdate moves to a new motor the motorFrequency gets updated:

```
                if (++motor == getMotorCount()) {
                    motor = 0;
                }
                motorFrequency[motor] = erpmToHz * filteredMotorErpm[motor];
```
This updated motorFrequency then does get used for all filters for that motor even if it takes more than one invocation of rpmFilterUpdate to update them. That's done to minimize cpu cycles spent. Every filter is updated at least once per ms and we have four motors, so there's no significant delay here (250 uS or one cycle max).

Reply:
The structure of rpmFilterUpdate might seem a bit weird. It's done the way it is so that it allows arbitrary partitioning of filter updates over loop cycles.

Reply:
Actually using default settings if filterUpdatesPerIteration were 1 you'd have 4 loop iterations before the motor changes since the dterm only has one harmonic by default. So in that case instead of calling ``motorFrequency[motor] = erpmToHz * filteredMotorErpm[motor]`` once we'd call it four times if that were moved up.

But the key is that filterUpdatesPerIteration is calculated at init so that all filters are updated in 1ms. Since all filters belong to 4 motors that leaves 250uS for each motor. Which is just not a significant delay for the rpm data which is very noisy anyhow.

The frequency from harmonic calculation is only up there because it needs to happen per iteration anyway.

Reply:
One thing to decide is if it makes sense to update the rpmFilter in the pid loop vs the gyro loop. I did this with an eye on https://github.com/betaflight/betaflight/issues/7326. If gyro runs much higher than pids it's better to run more stuff at pid frequency to have less function call and loop overhead.